go/libraries/utils/filesys/lock.go

```diff
@@ -0,0 +1,91 @@
+package filesys
+
+import (
+	"sync/atomic"
+
+	"github.com/juju/fslock"
+	"github.com/pkg/errors"
+)
+
+const originalStateValue int32 = 0
+const newStateValue int32 = 1
+
+var errLockUnlock = errors.New("unable to unlock the lock")
+
+type FilesysLock interface {
+	TryLock() (bool, error)
+	Unlock() error
+}
+
+func CreateFilesysLock(fs Filesys, filename string) FilesysLock {
+	switch fs.(type) {
+	case *InMemFS:
+		return NewInMemFileLock(fs)
+	case *localFS:
+		return NewLocalFileLock(fs, filename)
+	default:
+		panic("Unsupported file system")
+	}
+}
+
+type InMemFileLock struct {
+	state int32
+}
+
+func NewInMemFileLock(fs Filesys) *InMemFileLock {
+	return &InMemFileLock{originalStateValue}
+}
+
+func (memLock *InMemFileLock) TryLock() (bool, error) {
+	if atomic.CompareAndSwapInt32(&memLock.state, originalStateValue, newStateValue) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (memLock *InMemFileLock) Unlock() error {
+	if memLock.state == originalStateValue {
+		return nil
+	}
+
+	// old := newStateValue
+	new := atomic.AddInt32(&memLock.state, -newStateValue)
+
+	// if atomic.CompareAndSwapInt32(&memLock.state, old, new) {
+	// 	return nil
+	// }
+
+	if new != 0 {
+		return errLockUnlock
+	}
+
+	return nil
+}
+
```

Comment:
Looked at the implementation of `Mutex` lock and unlock methods to see how to use this. I think I'm correct or close. A test would confirm this possibly incorrect suspicion of this being close.